test/hotspot/gtest/nmt/test_nmt_buffer_overflow_detection.cpp

```diff
@@ -50,101 +50,18 @@
-DEFINE_TEST(test_overwrite_back_long_aligned_distance, "footer canary broken")
-static void test_overwrite_back_long_unaligned_distance() { test_overwrite_back_long(0x2001); }
-DEFINE_TEST(test_overwrite_back_long_unaligned_distance, "footer canary broken")
-
-///////
-
-static void test_double_free() {
-  address p = (address) os::malloc(1, mtTest);
-  os::free(p);
-  // Now a double free. Note that this is susceptible to concurrency issues should
-  // a concurrent thread have done a malloc and gotten the same address after the
-  // first free. To decrease chance of this happening, we repeat the double free
-  // several times.
-  for (int i = 0; i < 100; i ++) {
-    os::free(p);
-  }
-}
-
-// What assertion message we will see depends on whether the VM wipes the memory-to-be-freed
-// on the first free(), and whether the libc uses the freed memory to store bookkeeping information.
-// If the death marker in the header is still intact after the first free, we will recognize this as
-// double free; if it got wiped, we should at least see a broken header canary.
-// The message would be either
-// - "header canary broken" or
-// - "header canary dead (double free?)".
-// However, since gtest regex expressions do not support unions (a|b), I search for a reasonable
-// subset here.
-DEFINE_TEST(test_double_free, "header canary")
-
-///////
-
 static void test_invalid_block_address() {
   // very low, like the result of an overflow or of accessing a null this pointer
   os::free((void*)0x100);
 }
 DEFINE_TEST(test_invalid_block_address, "invalid block address")
 
-///////
-
 static void test_unaliged_block_address() {
   address p = (address) os::malloc(1, mtTest);
   os::free(p + 6);
 }
 DEFINE_TEST(test_unaliged_block_address, "block address is unaligned");
 
-///////
-
-// Test that we notice block corruption on realloc too
-static void test_corruption_on_realloc(size_t s1, size_t s2) {
-  address p1 = (address) os::malloc(s1, mtTest);
-  *(p1 + s1) = 'a';
-  address p2 = (address) os::realloc(p1, s2, mtTest);
-
-  // Still here?
-  tty->print_cr("NMT did not detect corruption on os::realloc?");
-  // Note: don't use ASSERT here, that does not work as expected in death tests. Just
-  // let the test run its course, it should notice something is amiss.
-}
-static void test_corruption_on_realloc_growing()    { test_corruption_on_realloc(0x10, 0x11); }
-DEFINE_TEST(test_corruption_on_realloc_growing, COMMON_NMT_HEAP_CORRUPTION_MESSAGE_PREFIX);
-static void test_corruption_on_realloc_shrinking()  { test_corruption_on_realloc(0x11, 0x10); }
-DEFINE_TEST(test_corruption_on_realloc_shrinking, COMMON_NMT_HEAP_CORRUPTION_MESSAGE_PREFIX);
-
-///////
-
 // realloc is the trickiest of the bunch. Test that realloc works and correctly takes over
 // NMT header and footer to the resized block. We just test that nothing crashes - if the
 // header/footer get corrupted, NMT heap corruption checker will trigger alert on os::free()).
```

Comment:
After removing canary, can NMT still check memory corruption?

Reply:
No, not anymore. The canaries are how we perform memory corruption checks. As mentioned in the PR description, I believe that since we have introduced UBSan and ASan it is no longer necessary for NMT to be able to detect memory corruption.

Reply:
I would like to caution that this would trash the ability to check for corruption for platforms that don't support ubsan and asan. Windows, with the VC compiler, comes to mind (Actually, even the Windows/gcc Port cannot use ubsan and asan either, since it hasn't been properly ported for Windows in general)

Reply:
clang and its many variants (clang\-cl, clang in MinGW mode, etc) are not supported as compilers for the Windows JDK. The only compiler that the build system allows for Windows is VC. I do have a port that supports gcc (Which unfortunately also does not have access to ubsan or asan), but have not started work on clang because it was a bit above my capabilities

Reply:
Interesting. I'm personally not aware of why using a different C++-compiler on a particular platform would be a problem.

Reply:
Since ASAN is supported by VC on Windows platforms, we can still be safe to remove the memory corruption checking by NMT. ASAN is already available in JDK build, so available for the Windows platform.